### PR TITLE
ReduceToPlane

### PR DESCRIPTION
**Pull request by WeiqunZhang**
* Add a function template for reducing MultiFab data to a plane.

* Add KeyValuePair as an alias to ValLocPair.

* Add more versions of MPI reduce functions for KeyValuePair/ValLocPair.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [x] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
@marchdf

**Review by jmsexton03 — APPROVED** [review]

File: Src/Base/AMReX_MultiFabUtil.H [hunk 1]
```diff
@@ -246,6 +246,75 @@ namespace amrex
...
+     * \brief Reduce FabArray/MultiFab data to a plane.
+     *
+     * This function takes a FabArray/MultiFab and reduces its data to a
+     * plane.  The return data are stored in a BaseFab with only one cell in
+     * the normal direction of the plane.  The index range of the BaseFab in
+     * the other directions is the same as the provided domain Box.  If data
+     * do not exist along a certain line, the value is set to the minimum,
+     * maximum and zero, for reduce max, min and sum, respectively.  The
+     * reduction is local and the user may need to do MPI communication
+     * afterwrads if needed.
```
Comment (marchdf) [inline Src/Base/AMReX_MultiFabUtil.H]: typo in `afterwards`

Comment (WeiqunZhang) [inline Src/Base/AMReX_MultiFabUtil.H]:
```suggestion
     * afterwards if needed.
```

File: Src/Base/AMReX_MultiFabUtil.H [hunk 2]
```diff
@@ -246,6 +246,75 @@ namespace amrex
...
+         auto const& domain_box = geom.Domain().surroundingNodes(); // nodal data
+         auto const& ma = mf.const_arrays();
+         auto rr = ReduceToPlane<ReduceOpMax,KeyValuePair<Real,int>>
+             (dir, domain_box, mf,
+              [=] AMREX_GPU_DEVICE (int box_no, int i, int j, int k)
+                  -> KeyValuePair<Real,int>
+             {
+                 return {ma[box_no](i,j,k), i};
+             });
+         ParallelReduce::Max(rr.dataPtr(), r.size(), root,
```
Comment (marchdf) [inline Src/Base/AMReX_MultiFabUtil.H]: typo: should be `rr.size()`

File: Src/Base/AMReX_MultiFabUtil.H [hunk 3]
```diff
@@ -246,6 +246,75 @@ namespace amrex
...
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_MultiFabUtil.H]:
```suggestion
         ParallelReduce::Max(rr.dataPtr(), rr.size(), root,
```

**Pull request merged**